base-notebook/fix-passwd-entry

```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+# By default the image would run as user ID 1000. If the user is run as
+# an arbitrarily assigned user ID by the container platform it will be
+# much higher and the lack of entries in /etc/passwd and /etc/group can
+# cause failure of any Python code which doesn't tolerate having no
+# entries for uid/gid. Assume that if running with larger uid that we
+# need to add entries into the respective files for that user. Have the
+# check be greater than 100000 to allow room in case custom images
+# derived from this image were adding additional explict users/groups of
+# their own.
+
+NB_UID=`id -u`
+NB_GID=`id -g`
+
+if [ $NB_GID -eq 0 -a $NB_UID -ge 100000 ]; then
+    cat /etc/passwd | sed -e "s/^jovyan:/nayvoj:/" > /tmp/passwd
+    echo "jovyan:x:$NB_UID:$NB_GID:,,,:/home/jovyan:/bin/bash" >> /tmp/passwd
+    cat /tmp/passwd > /etc/passwd
+    rm /tmp/passwd
+
+    id -G | grep -q -w $NB_UID; STATUS=$?
+    if [ $STATUS -eq 0 ]; then
+        echo "jovyan:x:$NB_UID:" >> /etc/group
+    fi
```

Comment:
We have similar logic to this in start.sh, and it currently enables:

- setting user id
- setting user name
- granting sudo access

The start.sh logic requires launching the image as the root *user* rather than the root group. Effectively, we have a fork of two different behaviors with the same goal - one dedicated for openshift/some kubernetes deployments (this one) and one for most others (in start.sh). They accomplish the same thing in different ways and in different code paths. Perhaps we should reconcile them. It would probably be nice if we could accomplish the existing root-requiring logic in start.sh via root *group* rather than root *user*. Would you like to take a stab at that?

Reply:
The problem still gets back to fact that JupyterHub only runs ``jupyterhub-singleuser`` and not ``start-singleuser.sh``. Because of my use case for JupyterHub that isn't a complete stumbling block though as it is already necessary to configure ``KubeSpawner`` to set supplemental group of ``users`` on the container since ``docker-stacks`` images aren't setup to use group ``root`` for everything, but rely on groups ``users`` when user ID is not 1000. Since already doing that, no drama to also override the startup command to be ``start-singleuser.sh``. Since it internally uses ``start.sh`` to run ``jupyterhub-singleuser`` then integrating it with ``start.sh`` is probably okay.

I can't see that this will change the need to make ``passwd`` and ``group`` files writable to ``root`` group. The only way around that is messy and requires compiled C programs that are ``setuid`` so can become effective user of ``root``, but not real user as kernel setuid() call blocked, such that can update files in that C program. Am not confident that isn't blocked by some configurations for running containers.

Even if make ``passwd`` and ``group`` writable, my recollection is that commands like ``usermod`` and ``groupmod`` don't work as that change is not enough. When I tested that it may have been on CentOS though so possible that Debian is different. So chance that have to still make changes to ``passwd`` and ``group`` manually. I will do some more testing and see what can be done using ``start.sh``.